source/client-side-encryption/client-side-encryption.rst

```diff
@@ -708,20 +708,44 @@ ClientEncryption
 
       // Creates a new key document and inserts into the key vault collection.
       // Returns the \_id of the created document as a UUID (BSON binary subtype 4).
-      createKey(kmsProvider: String, opts: Optional<DataKeyOpts>): Binary;
+      createKey(kmsProvider: String, opts: Optional<DataKeyOpts>): UUID;
 
       // An alias function equivalent to createKey.
-      createDataKey(kmsProvider: String, opts: Optional<DataKeyOpts>): Binary;
+      createDataKey(kmsProvider: String, opts: Optional<DataKeyOpts>): UUID;
 
       // Decrypts multiple data keys and (re-)encrypts them with a new masterKey, or with their current masterKey if a new one is not given.
       // Returns a RewrapManyDataKeyResult.
       rewrapManyDataKey(filter: Document, opts: Optional<RewrapManyDataKeyOpts>): RewrapManyDataKeyResult;
 
+      // Removes the key document with the given \_id from the key vault collection.
+      // Returns the result of the internal deleteOne() operation on the key vault collection.
+      deleteKey(id: UUID): DeleteResult;
+
+      // Finds a single key document with the given UUID.
+      // Returns the result of the internal find() operation on the key vault collection.
+      getKey(id: UUID): Iterable<Document>;
+
+      // Finds all documents in the key vault collection.
+      // Returns the result of the internal find() operation on the key vault collection.
+      getKeys(): Iterable<Document>;
+
+      // Adds a keyAltName to the keyAltNames array of the key document in the key vault collection with the given \_id.
+      // Returns the previous version of the key document.
+      addKeyAlternateName(id: UUID, keyAltName: String): Optional<Document>;
+
+      // Removes a keyAltName from the keyAltNames array of the key document in the key vault collection with the given \_id.
+      // Returns the previous version of the key document.
+      removeKeyAlternateName(id: UUID, keyAltName: String): Optional<Document>;
+
+      // Returns a key document in the key vault collection with the given \_id.
+      getKeyByAltName(keyAltName: String): Iterable<Document>;
```

Comment:
Can multiple keys share the same `keyAltName`? I assume not, as you're creating a partial, unique index on this field. In that case, wouldn't it make sense for this to return an `Optional<Document>`?

Reply:
> `Array<Document>` (for `getKeys()`)

I have no objection to using `Iterable<Document>` there, as there is precedent for doing so in other specs. Drivers are free to expose a cursor or return any iterable type.

I'm only focused on `getKeyByAltName`. This creates a more complicated API for any consumer of this method, since they need to iterate just to access a potential result. In the case of libmongoc, they might also have to free the cursor (in addition to anything they might copy out of it).

Consider the `findAndModify` methods in the CRUD spec. Implementation details aside (`findAndModify` doesn't use a cursor), consumers of those methods expect to either get a single document result or nothing at all. Returning an `Optional<Document>` here seems like something [users would most expect](https://en.wikipedia.org/wiki/Principle_of_least_astonishment).

Reply:
I am convinced and in agreement. Thank you for pointing to `findAndModify` (i.e. `findOneAndUpdate`) as reference. I will change the return type of functions that return "zero or one" documents (both `getKey()` and `getKeyByAltName()` for consistency) from `Iterable<Document>` to `Optional<Document>` as suggested.